pupil_src/shared_modules/video_capture/hmd_streaming.py

```diff
@@ -35,23 +34,116 @@ def __init__(self, buffer, timestamp, index, width, height):
         self.yuv_buffer = None
         self.jpeg_buffer = None
 
+    def interpret_buffer(self, buffer, width, height) -> np.ndarray:
+        return np.fromstring(buffer, dtype=np.uint8).reshape(height, width, self.depth)
+
+    @property
+    @abc.abstractmethod
+    def depth(self) -> int:
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def gray(self) -> np.ndarray:  # dtype uint8, shape (height, width)
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def bgr(self) -> np.ndarray:
+        # dtype uint8, shape (height, width, 3), memory needs to be allocated contiguous
+        raise NotImplementedError
+
+    @property
+    def img(self) -> np.ndarray:  # equivalent for bgr; kept for legacy reasons
+        return self.bgr
 
-FRAME_CLASS_BY_FORMAT = {"rgb": RGBFrame}
+
+class BGRFrame(Uint8BufferFrame):
+    @property
+    def depth(self) -> int:
+        return 3
+
+    @property
+    def bgr(self) -> np.ndarray:
+        return self._buffer
+
+    @property
+    def gray(self):
+        try:
+            return self._gray
+        except AttributeError:
+            self._gray = np.mean(self._buffer, axis=-1).astype(self._buffer.dtype)
+            return self._gray
+
+
+class RGBFrame(BGRFrame):
+    @property
+    def depth(self) -> int:
+        return 3
+
+    @property
+    def bgr(self) -> np.ndarray:
+        try:
+            return self._bgr
+        except AttributeError:
+            self._bgr = np.ascontiguousarray(np.flip(self._buffer, (0, 2)))
+            return self._bgr
+
+    @property
+    def gray(self):
+        try:
+            return self._gray
+        except AttributeError:
+            self._gray = np.mean(self._buffer, axis=-1).astype(self._buffer.dtype)
+            return self._gray
+
+
+class GrayFrame(BGRFrame):
```

Comment:
ditto about superclass

Reply:
That one should change
```suggestion
class GrayFrame(Uint8BufferFrame):
```